pkg/server/sandbox_portforward_unix.go

```diff
@@ -62,48 +61,58 @@ func (c *criService) portForward(ctx context.Context, id string, port int32, str
-		// from stream finishes. Unfortunately, if you have a client like telnet connected
-		// via port forwarding, as long as the user's telnet client is connected to the user's
-		// local listener that port forwarding sets up, the telnet session never exits. This
-		// means that even if socat has finished running, command.Run() won't ever return
-		// (because the client still has the connection and stream open).
-		//
-		// The work around is to use StdinPipe(), as Wait() (called by Run()) closes the pipe
-		// when the command (socat) exits.
-		in, err := cmd.StdinPipe()
+		// TODO: hardcoded to tcp4 because localhost resolves to ::1 by default if the system has IPv6 enabled.
+		// Theoretically happy eyeballs will try IPv6 first and fallback to IPv4
+		// but resolving localhost doesn't seem to return and IPv4 address, thus failing the connection.
+		conn, err := net.Dial("tcp4", fmt.Sprintf("localhost:%d", port))
 		if err != nil {
-			return errors.Wrap(err, "failed to create stdin pipe")
+			return errors.Wrapf(err, "failed to dial %d", port)
 		}
+
+		errCh := make(chan error, 2)
+		// Copy from the the namespace port connection to the client stream
 		go func() {
-			if _, err := io.Copy(in, stream); err != nil {
-				logrus.WithError(err).Errorf("Failed to copy port forward input for %q port %d", id, port)
-			}
-			in.Close()
-			logrus.Debugf("Finish copying port forward input for %q port %d", id, port)
+			// tries to shutdown only its destination stream, giving it a chance to terminate gracefully
+			defer stream.Close()
+			log.G(ctx).Debugf("PortForward copying data from namespace %q port %d to the client stream", id, port)
+			_, err := io.Copy(stream, conn)
+			errCh <- err
 		}()
 
-		if err := cmd.Run(); err != nil {
-			return errors.Errorf("socat command returns error: %v, stderr: %q", err, stderr.String())
+		// Copy from the client stream to the namespace port connection
+		go func() {
+			// tries to shutdown only its destination connection, giving it a chance to terminate gracefully
+			defer conn.Close()
+			log.G(ctx).Debugf("PortForward copying data from client stream to namespace %q port %d", id, port)
+			_, err := io.Copy(conn, stream)
+			errCh <- err
+		}()
+
+		// Wait until the first error is returned and return it
+		// because is the cause the port forwarding stops running
+		errFwd := <-errCh
+
+		// give a chance to terminate gracefully or timeout
+		// 0.5s is the default timeout used in socat
+		// https://linux.die.net/man/1/socat
+		timeout := time.Duration(500) * time.Millisecond
+		select {
+		case e := <-errCh:
+			if e != nil {
+				// This error may be caused because we are shutting down the connection, only log it
+				log.G(ctx).Debugf("PortForward closed the connection with error %v in network namespace %q port %d", e, id, port)
+			} else {
+				log.G(ctx).Debugf("PortForward closed the connection gracefully in network namespace %q port %d", id, port)
+			}
+		case <-time.After(timeout):
```

Comment:
how about using var `ctx` for timeout?

Reply:
or both.. in case ctx isn't timing out for some odd reason waiting on the streams to close :)

Reply:
then I think that the context cancellation should be done when we are blocked waiting for the first error, right?

Reply:
5e7b3e4e35e8c4a94dd967681b06e2a2b61ff4c3